libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/PlannedRouteRefreshController.kt

```diff
@@ -0,0 +1,138 @@
+package com.mapbox.navigation.core.routerefresh
+
+import androidx.annotation.VisibleForTesting
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.base.route.RouteRefreshOptions
+import com.mapbox.navigation.utils.internal.logI
+import kotlinx.coroutines.CoroutineScope
+
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
+internal class PlannedRouteRefreshController @VisibleForTesting constructor(
+    private val routeRefresherExecutor: RouteRefresherExecutor,
+    private val routeRefreshOptions: RouteRefreshOptions,
+    private val stateHolder: RouteRefreshStateHolder,
+    private val listener: RouteRefresherListener,
+    private val cancellableHandler: CancellableHandler,
+    private val retryStrategy: RetryRouteRefreshStrategy,
+) : Pausable {
+
+    constructor(
+        routeRefresherExecutor: RouteRefresherExecutor,
+        routeRefreshOptions: RouteRefreshOptions,
+        stateHolder: RouteRefreshStateHolder,
+        scope: CoroutineScope,
+        listener: RouteRefresherListener,
+    ) : this(
+        routeRefresherExecutor,
+        routeRefreshOptions,
+        stateHolder,
+        listener,
+        CancellableHandler(scope),
+        RetryRouteRefreshStrategy(maxRetryCount = 2)
+    )
+
+    private var routesToRefresh: List<NavigationRoute>? = null
+
+    fun startRoutesRefreshing(routes: List<NavigationRoute>) {
+        cancellableHandler.cancelAll()
+        routesToRefresh = null
+        if (routes.isEmpty()) {
+            logI("Routes are empty", RouteRefreshLog.LOG_CATEGORY)
+            stateHolder.reset()
+            return
+        }
+        val routesValidationResults = routes.map { RouteRefreshValidator.validateRoute(it) }
+        if (
+            routesValidationResults.any { it is RouteRefreshValidator.RouteValidationResult.Valid }
+        ) {
+            routesToRefresh = routes
+            scheduleNewUpdate(routes)
+        } else {
+            val message =
+                RouteRefreshValidator.joinValidationErrorMessages(
+                    routesValidationResults.mapIndexed { index, routeValidationResult ->
+                        routeValidationResult to routes[index]
+                    }
+                )
+            val logMessage = "No routes which could be refreshed. $message"
+            logI(logMessage, RouteRefreshLog.LOG_CATEGORY)
+            stateHolder.onFailure(logMessage)
+            stateHolder.reset()
+        }
+    }
+
+    override fun pause() {
+        cancellableHandler.cancelAll()
+    }
+
+    override fun resume() {
+        routesToRefresh?.let {
+            if (retryStrategy.shouldRetry()) {
+                scheduleUpdateRetry(it)
+            }
+        }
+    }
```

Comment:
It's a bit complicated. I don't like it but I decided not to try to simplify everything perfectly in one step but just to start with something.
Pause will not cause receiving CANCELLED state. Therefore we don't need to dispatch STARTED in resume.
It **was** something like that:
Assume we need 3 attempts to refresh routes. It will be: STARTED -> attempt1 -> attempt2 -> attempt3 -> SUCCESS/FAILED and not STARTED -> attempt1 -> FAILED -> STARTED -> attempt2 -> FAILED -> STARTED -> attempt3 -> SUCCESS/FAILED.
Now `pause` will just cancel attempt2 if it is currently planned and `resume` will say: "resume the attempts starting from attempt2". In order to keep it as it was, `pause` and `resume` do not affect states. 
Does this make sense?

Reply:
I still don't fully understand. `pause` is invoked when an on-demand refresh is executed. If that refresh fails, we'll `resume` without dispatching on-start, so we have:
```
STARTED -> attempt1 -> on-demand (`pause`) -> FAILED -> (`resume` without STARTED) -> SUCCESS/FAILED
```
don't we? We're missing the start for the resumed attempt.

Reply:
Oh, I got confused. `pause` does dispatch the `CANCELLED` state. You are right, `STARTED` should be dispatched here. It was not caught in tests because now every time we do the resume the next moment we actually restart planned route refreshes (because we get new REFRESHED routes, more info here: https://github.com/mapbox/mapbox-navigation-android/pull/6610#discussion_r1032371777) so it doesn't really matter.